### ISSUE TITLE
Updating specific indexes

### ISSUE THREAD

**Issue by RubenSoh** [npgall/cqengine#258]
Hello Mr. Gallagher,

I know it's already been mentioned in #210, but that's been closed and I don't think the issue has been done justice. You state that it's difficult for CQEngine to know what attributes might have updated as a result of a field change, but if you know ahead of time what attributes/indexes are going to be affected, you can call some sort of `updateIndexesOnAttribute` method or something

For (a very simplistic) example:

```java
class Car {

    String description = "my car is dirty";

    public static final Attribute<Car, Boolean> IS_DIRTY = 
        attribute("is_dirty", car -> car.description.contains("dirty"));

}
```

Now say we want to change `description` rapidly without copying and re-instantiating or even re-indexing all of `Car`'s fields, as `Car` might be quite a large object. You possibly don't even need the index to always be up-to-date. 

All of the above scenarios could be flexibly covered if you had some method to update specific indexes on an object, which might look something like follows

```java
Car myCar = new Car();
IndexedCollection<Car> cars = new ConcurrentIndexedCollection<>();
cars.addIndex(Car.IS_DIRTY);
cars.add(myCar);
car.description = "the car is squeaky clean";
// Instead of calling addObject / removeObject or creating a new Car...
cars.updateObjectOnIndex(myCar, Car.IS_DIRTY);
```

Would you be open to adding such a feature or merging a PR with it? Are there any complications that might prevent you from doing so?

Regards and thanks for your time.

**Comment by npgall** [comment]
Yes I think there may be a number of complications (I think I mentioned some in #210).

My main concern is that it could be quite a _"sharp edge"_ if we were to introduce a feature like this. (By _sharp edge_, I mean a feature that is prone to being used incorrectly, and when it is used incorrectly, it can cause application failures.) Some complications which might arise are as follows...

**1. We can't easily provide an API to modify an attribute value**
As I mentioned in #210, CQEngine attributes are basically just functions, which only happen to be implemented _most of the time_, to produce their data by reading fields in objects. But as they really are functions, any given attribute can actually read from multiple fields (or external data sources) to compute its data. This means it's not possible to know how underlying fields should be updated, in order to change the value of a specific attribute.

Thus, it's difficult to provide an API like this (1A):

```java
collection.setAttributeValue(car, Car.CAR_AGE, newValue); // e.g. newValue == 10
```
or (1B):
```java
collection.updateAttributeValue(car, Car.CAR_AGE, oldValue, newValue); // e.g. oldValue == 5, newValue == 10
```

**2. Directly mutating fields in objects can leave stale entries in indexes**
In your example above, you are mutating the `car.description` field _after_ you have added the Car to the collection. You have an attribute IS_DIRTY which reads that field, and you already have built an index on the attribute. Ordinarily mutating a field in that way is quite risky, because it can leave stale entries in indexes.

For example, if an API like this was called _after_ the description field had _already_ been modified...
```java
collection.updateIndexesOn(Car.IS_DIRTY);
```
...then CQEngine would no longer be able to locate the old entry for that Car in the index(es). How would it know if the car previously had been dirty or not? (Ok, in this case it's a Boolean attribute and theoretically it could scan the index, checking for entries for both possible values - but that's not feasible for attributes which have a large number of possible values.)

Generally, in order to safely be able to update indexes, CQEngine needs to know both the _old_ and the _new_ value for the attribute; but the old value is lost whenever a field is mutated by the application.

This is why it's recommended to use object versioning instead. That is, to store *immutable* objects in the collection, and to _swap_ an old _version_ of an object with a new version of the same object instead - using the method [IndexedCollection.update()](http://htmlpreview.github.io/?http://raw.githubusercontent.com/npgall/cqengine/master/documentation/javadoc/apidocs/com/googlecode/cqengine/IndexedCollection.html#update(java.lang.Iterable,java.lang.Iterable)). `TransactionalIndexedCollection` can do that [atomically](http://htmlpreview.github.io/?http://raw.githubusercontent.com/npgall/cqengine/master/documentation/javadoc/apidocs/com/googlecode/cqengine/TransactionalIndexedCollection.html#update(java.lang.Iterable,java.lang.Iterable)).

You might also ask then why CQEngine can't provide an API to change an attribute's value from oldValue to newValue - but that is already discussed in 1B above.

**3. We really need a dependency graph to mutate fields**
As I mentioned in (1) above, it's not easy/feasible to allow _attribute_ values to be changed.

However, a natural question then is, what about providing an API to allow _field_ values to be changed instead? 

This theoretically would be possible. However, then we would need to maintain some kind of dependency graph, to enable CQEngine to determine which attributes read data from whichever field was to be modified, in order to determine which indexes were built on those attributes and so would need to be updated.

Currently, the only place where we can determine the fields from which an attribute reads its data, is by looking at the source code (or bytecode) of the attribute. Some kind of automated analysis may be possible, but it could be pretty complicated.

Hope that makes sense! I'm open to suggestions, but I think this stuff could get really complicated. And I think the approach of keeping objects immutable, and using [object versioning/MVCC](http://htmlpreview.github.io/?http://raw.githubusercontent.com/npgall/cqengine/master/documentation/javadoc/apidocs/com/googlecode/cqengine/TransactionalIndexedCollection.html) is generally much easier to work with.

**Comment by RubenSoh** [comment]
Hello, I understand the difficulty/impossibility of setting attributes as they are computed from (hopefully pure) functions

I have written a private (and somewhat messy) implementation of solution 2, in which stale indexes exist. While understand this is not the most coherent API, it is importantly applicable to use cases such as video games or other realtime applications where data mutation occurs rapidly, and immutable copying of objects is simply too slow and inefficient. 

To represent an index that might be possibly stale, I create the `MutableIndex<O>` interface, which supports an "updateAll" method in addition to an "addAll" and "removeAll" method. This relies on the objects having an immutable hashcode so that references can be always safely removed and added to hashmap. 

For example: on my implementation of 'MutableHashIndex'

    @Override
    public void updateAll(ObjectSet<O> objectSet, QueryOptions queryOptions) {
        try {
            for (O object : objectSet) {
                Iterable<A> oldAttributeValues = reverseMap.get(object);
                for (A oldAttributeValue : oldAttributeValues) {
                    StoredResultSet<O> oldAttribute = super.indexMap.get(oldAttributeValue);
                    oldAttribute.remove(object);
                }

                Iterable<A> newAttributeValues = getAttribute().getValues(object, queryOptions);
                for (A newAttributeValue : newAttributeValues) {
                    StoredResultSet<O> newAttribute = super.indexMap.computeIfAbsent(newAttributeValue, k -> valueSetFactory.create());
                    newAttribute.add(object);
                }
                reverseMap.put(object, newAttributeValues);
            }
        } finally {
            objectSet.close();
        }
    }

Thus, you can manually recalculate the index for certain objects when they're mutated and the index needs to be recalculated.

However, I agree that to use this properly requires foresight (a 'sharp edge', as you say) to correctly update indexes before they become stale. However, it is necessary for the use case in which the values contained within are updated hundreds of times per second. As a basic example, consider a player object. It would be nice to construct some type of spatial index based on their coordinates, and recreating said object hundreds of times and also recalculating indexes that don't need to be recalculated is not performant in the slightest. This is why I propose allowing mutation, but it's a specific use-case and understandable if you consider this out of scope of this library.
